### PR TITLE
Subscription System

### LINKED ISSUE #107: Manage Smart Slice stage access depending on user status
Interaction with the Smart Slice stage should be blocked unless, iff, the user is logged in and has an active subscription.

### PR DESCRIPTION
**Pull request by jake-kirkland-teton**
*Now checks for subscription statuses and counts for optimizations/validations.

*Displays popups to user to indicate expired subscription/trial

*Code is in place for when the user has specific number of validates/optimizes

-Only waiting on the last tests for trial expired system, then it should be good to merge.

Closes #107

### CODE REVIEW

**Comment by dougkenik** [comment]
@jake-kirkland-teton & @bradymadams - did you find an easy way to test this? I'm trying to figure out how to change my status (subscription / trial) and expiration

**Comment by jake-kirkland-teton** [comment]
Sent you a Slack.

File: SmartSlicePlugin/SmartSliceCloudConnector.py [hunk 1]
```diff
@@ -359,14 +361,80 @@ def _createTokenFile(self):
...
             json.dump(self._token, token_file)
 
+    def getProductUsesLeft(self, product):
+        apiCode, apiResult = self._client.smartslice_subscription()
+        notification_message = Message(lifetime=0)
+
+        if apiCode != 200:
+            self._handleThorErrors(apiCode, apiResult)
+            return 0
+        else:
```
Comment (bradymadams) [inline SmartSlicePlugin/SmartSliceCloudConnector.py:371]: The `else` isn't necessary since the above `if` will return if True.

Comment (jake-kirkland-teton) [reply SmartSlicePlugin/SmartSliceCloudConnector.py:371]: That it is not. Good point. I will get that adjusted. Thanks!

**Comment by jake-kirkland-teton** [comment]
Hmmm. I see what you mean. Let me think about it. I feel like there should be a way to achieve that without an enormous amount of effort.

**Comment by dougkenik** [comment]
Ok. There were a few bugs I found when adding/removing loads and editing values during a run. Should be good to take another look

**Comment by dougkenik** [comment]
Some thoughts on UX:

1. Can we make the hyperlink text Bold? The call to action is pretty weak from a user perspective
2. I think this dialog is the perfect reason to create a custom component for our own warnings. The message needs to be in the middle of the screen with a strong call to action. As of now, it's easy to miss. Thoughts?

**Comment by jake-kirkland-teton** [comment]
I agree. The message is not ideal in any way. It shows at the bottom and is not pressing. I can bold the text however.

I think we do need to look into creating our own messaging system. This is functional, but not ideal. So while it can be our bandage for right now, it should me improved for the long run. Do we feel that a better messaging system is required for MVP is the question?

**Comment by dougkenik** [comment]
> I agree. The message is not ideal in any way. It shows at the bottom and is not pressing. I can bold the text however.
> 
> I think we do need to look into creating our own messaging system. This is functional, but not ideal. So while it can be our bandage for right now, it should me improved for the long run. Do we feel that a better messaging system is required for MVP is the question?

Let's bold the text for this ticket, and create a new ticket to migrate to our own warning/error message dialog. I'll prioritize that ticket in our backlog from there.

**Comment by colman-hartinger** [comment]
is there a good way to test this?

**Comment by jake-kirkland-teton** [comment]
Sorry, meant to send it to you too when I sent it to Doug. Check slack.

File: SmartSlicePlugin/SmartSliceCloudConnector.py [hunk 2]
```diff
@@ -359,14 +361,80 @@ def _createTokenFile(self):
...
+                )
+
+            else:
+                notification_message.setText(
+                    i18n_catalog.i18nc("@info:status", "You do not have a subscription! Please subscribe to submit jobs.")
+                )
+                notification_message.addAction(
+                    action_id="subscribe_link",
+                    name="Manage Subscription",
+                    icon="",
```
Comment (colman-hartinger) [inline SmartSlicePlugin/SmartSliceCloudConnector.py:415]: can you just remove the `icon = ""` ? doesnt seem to have any effect

Comment (jake-kirkland-teton) [reply SmartSlicePlugin/SmartSliceCloudConnector.py]: It is a positional argument and required, so even though it is blank, it is required for this method, its not an optional sadly.

**Comment by jake-kirkland-teton** [comment]
OK. Merged and functioning better now. I think the current changes should be good to go. Let me know.

File: SmartSlicePlugin/SmartSliceCloudConnector.py [hunk 3]
```diff
@@ -520,6 +531,11 @@ class SmartSliceCloudConnector(QObject):
     debug_save_smartslice_package_preference = "smartslice/debug_save_smartslice_package"
     debug_save_smartslice_package_location = "smartslice/debug_save_smartslice_package_location"
 
+    subscriptionExpired = 0
+    trialExpired = 1
+    outOfUses = 2
+    noSubscription = 3
```
Comment (bradymadams) [inline SmartSlicePlugin/SmartSliceCloudConnector.py:537]: Let's put these in a proper `enum.Enum`.

Comment (jake-kirkland-teton) [reply SmartSlicePlugin/SmartSliceCloudConnector.py]: It has been done.

**Comment by bradymadams** [comment]
Will merge pending approval from @dougkenik @colman-hartinger

**Comment by dougkenik** [comment]
I think the font on the hyperlinks need to be bigger. The whole thing is easily missed. How about changing the font for that text to "medium_bold" in the Cura theme?

**Comment by jake-kirkland-teton** [comment]
Alright Doug, no luck on using the built in Cura themes, but I did get it to be in a bigger font.

**Comment by colman-hartinger** [comment]
I think the message for managing the subscriptions looks a little funky

msg

besides that the coed looks good to me though

**Pull request merged**